src/db.ts

```diff
@@ -0,0 +1,401 @@
+            action: result.action,
+            prID: result.pr_id,
+            githubIssue: result.github_issue,
+          });
+        })
+        .catch((err) => {
+          reject(err);
+        });
+    });
+  }
+
+  /**
+   * Initialize the database by enabling foreign keys and creating all necessary tables and indices if not present.
+   * @return {Promise<void>} a promise that resolves to nothing and rejects if the database couldn't be initialized.
+   */
+  async init(): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.db = new sqlite3.Database(this.dbPath, (err) => {
+        if (err) {
+          reject(err);
+          return;
+        }
+
+        const statements = ["PRAGMA foreign_keys = ON"].concat(createSQLStatements);
+        const doNextStatement = (index: number = 0) => {
+          this.run(statements[index])
+            .then(() => {
+              // On the last statement, resolve the outer promise.
+              if (index === statements.length - 1) {
+                resolve();
+                return;
+              }
+
+              doNextStatement((index += 1));
+            })
+            .catch((err) => {
+              reject(err);
+            });
+        };
+
+        doNextStatement();
+      });
+    });
+  }
+
+  /**
+   * Wrap the sqlite3.Database.run method as a promise.
+   * @param {string} sql the SQL statement.
+   * @param {Array<any>} params an array of parameters for prepared statements.
+   * @return {Promise<sqlite3.RunResult>} a promise of a sqlite3.RunResult and rejects if the run method failed.
+   */
+  private async run(sql: string, params: Array<any> = []): Promise<sqlite3.RunResult> {
+    const db = this.db;
+    return new Promise((resolve, reject) => {
+      db?.run(sql, params, function (err) {
+        if (err) {
+          reject(err);
+          return;
+        }
+
+        resolve(this);
+      });
+    });
+  }
+
+  /**
+   * Set the last handled pull-request (PR) for the forked repo, forked repo branch, and upstream repo.
+   *
+   * This signifies the starting point for discovering new merged upstream PRs.
+   * @param {Repo} repo the Repo object representing the forked repo.
+   * @param {Repo} upstreamRepo the Repo object representing the upstream repo.
+   * @param {string} branch the forked repo's branch.
+   * @param {number} upstreamPRID the upstream GitHub PR ID.
+   * @return {Promise<void>} a promise that resolves to nothing and rejects if the database statement failed.
+   */
+  async setLastHandledPR(repo: Repo, upstreamRepo: Repo, branch: string, upstreamPRID: number): Promise<void> {
+    return new Promise((resolve, reject) => {
+      const sql = `
+          INSERT INTO last_handled_prs (
+            repo_id, branch, upstream_repo_id, upstream_pr_id
+          )
+          VALUES(?, ?, ?, ?)
+          ON CONFLICT(repo_id, upstream_repo_id, branch)
+          DO UPDATE SET upstream_pr_id=excluded.upstream_pr_id
```

Comment:
If there are two pending PRs, say `7` and `12`, what will happen if `12` gets handled first, and then `7`? It seems like this would update the row to say that `7` was last handled, which is correct, but seems confusing later, since something might think that `12` still needs to be handled?